src/net/transport.rs

```diff
@@ -30,115 +36,146 @@ impl HttpTransport {
+            format!("expected {name} in request"),
+        )
+    })
+}
+
 #[async_trait]
 impl Transport for Arc<HttpTransport> {
-    type CommandStream = ReceiverStream<CommandEnvelope>;
+    type RecordsStream = Empty<Vec<u8>>; // TODO(server): resolve placeholder
 
     fn identity(&self) -> HelperIdentity {
-        self.id
+        self.identity
     }
 
-    async fn subscribe(&self, subscription: SubscriptionType) -> Self::CommandStream {
-        match subscription {
-            SubscriptionType::QueryManagement => ReceiverStream::new(
-                self.subscribe_receiver
-                    .lock()
-                    .unwrap()
-                    .take()
-                    .expect("subscribe should only be called once"),
-            ),
-            SubscriptionType::Query(query_id) => {
-                let mut ongoing_queries = self.ongoing_queries.lock().unwrap();
-                match ongoing_queries.entry(query_id) {
-                    Entry::Occupied(_) => {
-                        panic!("attempted to subscribe to commands for query id {}, but there is already a previous subscriber", query_id.as_ref())
-                    }
-                    Entry::Vacant(entry) => {
-                        let (tx, rx) = mpsc::channel(1);
-                        entry.insert(tx);
-                        ReceiverStream::new(rx)
-                    }
-                }
+    async fn send<
+        D: Stream<Item = Vec<u8>> + Send + 'static,
+        Q: QueryIdBinding,
+        S: StepBinding,
+        R: RouteParams<RouteId, Q, S>,
+    >(
+        &self,
+        dest: HelperIdentity,
+        route: R,
+        data: D,
+    ) -> Result<(), TransportError>
+    where
+        Option<QueryId>: From<Q>,
+        Option<Step>: From<S>,
+    {
+        let route_id = route.resource_identifier();
+        match route_id {
+            RouteId::Records => {
+                // TODO(600): These fallible extractions aren't really necessary.
+                let query_id = extract_route_param(route.query_id(), "query_id")?;
+                let step = extract_route_param(route.step(), "step")?;
+                let resp_future = self.clients[dest]
+                    .step(dest, query_id, &step, data)
+                    .await
+                    .map_err(|_e| {
+                        io::Error::new(io::ErrorKind::ConnectionAborted, "channel closed")
+                    })?;
+                tokio::spawn(async move {
+                    resp_future
+                        .map_err(Into::into)
+                        .and_then(MpcHelperClient::resp_ok)
+                        .await
+                        .expect("failed to stream records");
+                });
+                // TODO(600): We need to do something better than panic if there is an error sending the
+                // data. Note, also, that the caller of this function (`GatewayBase::get_sender`)
+                // currently panics on errors.
```

Comment:
A panic isn't good, but we can (and should) collect all of these tasks somewhere and poll them alongside the other work on the task.  Spawning them is probably fine, but maybe we can track them in a `FuturesUnordered` somewhere here that we can periodically on the server end (probably in the task we use to consume input, which is the one we'll use to return the answer to the report collector ?).

Reply:
I would propose to come back to this later (when we deal with #600), unless you really want to see it addressed in this PR?